### PR TITLE
Fetching PS/2 data more quickly

### PR DESCRIPTION
**Pull request by stefan-b-jakobsson**
This is the SMC firmware part implementing a quicker way to fetch PS/2 data.

It replaces the Wire library with custom code that handles the hardware based I2C transmissions.

For more details, goto:

https://github.com/X16Community/x16-rom/pull/253

### CODE REVIEW

**Comment by stefan-b-jakobsson** [comment]
Summary of changes.

WIRE LIBRARY
- The Arduino Wire library has been replaced with a custom I2C slave implementation "SmcWire"
- This was needed to get finer control over the I2C bus communication, especially to make it possible to make the SMC respond to multiple slave addresses
- The custom library uses the USI hardware registers for I2C communication, similar to the Wire library.
- The current address 0x42 is left unchanged and is intended to work as it does today
- The new addresses 0x43 and 0x44 returns the next key code and the next mouse packet. As the Kernal does not need to set a command offset over I2C before fetching key codes and mouse packets, the speed of transmission is greatly improved. If there are no data to fetch, addresses 0x43 och 0x44 NACK the request. Returning NACK instead of a zero value also improves performance a lot.

PS/2 INIT
- mouse.h that contained both mouse and keyboard init machines is split into setup_mouse.h and setup_keyboard.h
- The keyboard and mouse init machines have been cleaned up and reorganized to make the flow easier to follow
- The mouse initializtion now also sets resolution, which might have been a problem for some mice
- Internal state variables of the init machines are no longer exposed. Instead you access the state through functions such as mouseReset() and mouseGetId()

X16-SMC.INO CLEAN-UP
- The main file has been cleaned up a lot, mostly without changing any code
- At the top of the file, all includes, definitions and global variables are collected under specific headers, making it easier to find things
- The functions are grouped as follows to make the code easier to navigate: (1) Setup, (2) Main loop, (3) Power, NMI, Reset, (4) I2C, (5), (6) PS/2, (7) Bootloader, and (8) ISRs
- The I2C_Receive function has been reorganized to use a switch statement instead of multiple if statements. This reduces code size and should speed up execution somewhat. The I2C_Process function has been incorporated into I2C_Receive for the same reasons.
- The I2C_Send function has also been changed to use a switch statement instead of multiple ifs
- The I2C hard reboot command has been fixed. It probably failed as it used a delay() statement in code called from within I2C_Receive/I2C_Process that is part of an interrupt service handler. The delay function needs the timer interrupt to keep the time. The ATtiny does not by default support nested interrupts, and I assume that the execution just hanged. The fix is based on setting a hard reboot request flag in I2C_Receive. The main loop checks if the flag is set, and will do the hard reboot if requested. This is the same technique used for reset with Ctrl+Alt+Del.

**Comment by stefan-b-jakobsson** [comment]
The first implementation of the fast PS/2 data fetch was based on making the SMC respond to three different slave addresses: the original address, mostly for commands, a new address dedicated for key codes and a new address dedicated for mouse packets.

While this worked great with the ATtiny861, after discussions with Wavicle it became clear that this setup might be a problem if future boards were to use another MCU as SMC, for instance the ESP32. Most other MCU's don't give you the same kind of low level control over the I2C protocol as the ATtiny861, and it would be difficult to make them respond to more than two slave addresses. It is most future proof to keep the SMC responding to just one slave address.

This second implementation uses only one slave address but it will still speed up the transmission speed as much or even more than the first implementation.

It works as follows:

- A default SMC read operation has been implemented. If you read from the SMC without first setting a command offset, the value of the default read operation is returned.

- You may set what is the default read operation by calling command offset 0x40 (CMD_SET_DFLT_READ_OP).

- Three new command offset have been added to the SMC that are inteded to be used as default read operations:

- 0x41 CMD_GET_KEYCODE_FAST: Returns the next key code. If no data is available, the I2C request is NACKed/ignored

- 0x42 CMD_GET_MOUSE_MOV_FAST: Returns the next mouse packet (3 or 4 bytes). If no data is available, the I2C request is NACKed/ignored.

- 0x43 CMD_GET_PS2DATA_FAST: Returns both the next key code and the next mouse packet. The key code is first, and the mouse packet is second. If only one of them is available, the a zero is returned in place of the other value. If neither one is available, the I2C request is NACKed/ignored.

Corresponding changes have been made to the Kernal PR. The Kernal code uses CMD_KEYCODE_FAST after init. If the mouse is enabled, it uses CMD_GET_PS2DATA_FAST. It returns to CMD_GET_KEYCODE_FAST if the mouse is disabled.

Speed comparision to the first implementation:

- Mouse disabled: No difference

- Mouse enabled: If there is no key code or mouse packet to fetch, which is the most common scenario, only one I2C request is made for both keyboard and mouse. That will save some 70 us compared to the first implementation. If there is a key code but no mouse packet, there should be no speed difference to the first implementation. If there is a mouse packet but no key code, there should be no difference. Finally, if there is both a key code and a mouse packet, we should save 70 us compared to the first implementation.

And one last thing. If this PR is merged, I ask permission to be added to the list of contributors. I hope that is OK with everybody.

**Comment by stefan-b-jakobsson** [comment]
The latest additions to the code address a problem with some newly manufactured mice, at least the Perimice 201. Even though the Perimice 201 acknowledges the set sample rate command, it will ignore the setting and continue to send movement reports at 100 Hz.

As the Kernal fetches mouse movement reports at 60 Hz (once per VBLANK), the SMC mouse buffer quickly overflows, preventing the Perimice 201 from working.

The issue has been addressed in the following way:

- The SMC doesn't store incoming bytes directly in the output buffer, but keeps the current movement packet (3 or 4 bytes) separate
- When the SMC has received a complete packet it will try to update the previous packet in the buffer instead of adding a new packet to the buffer.
- The delta X and delta Y of the two packages are simply added together
- The wheel movements are also added together
- If adding the packets together would cause an overflow, the update is aborted, and the SMC will push the last received packet as is to the buffer
- The last received packet is also pushed to the buffer as is if the button state has changed.

The change will also be a bit more resilient to SMC mouse buffer overflow for mice sending reports at the requested 60 Hz, which is good.

My first implementation handled the 2's complement math on a bit level. Even if this was working fine, it made the code hard to understand and maintain. I have now changed the code to use c++ signed math. The drawback is that the size of the code grew some 100 bytes by doing this. As a side note, the c++ compiler isn't producing very efficient code, but it is what it is.

File: x16-smc.ino [hunk 1]
```diff
@@ -3,93 +3,132 @@
 // By: Kevin Williams - TexElec.com
 //     Michael Steil
 //     Joe Burks
+//     Stefan Jakobsson
```
Comment (jburks) [inline x16-smc.ino:6]: My favorite change of this PR!

**Review by jburks — APPROVED** [review]
This is a large, but necessary, SMC update. It fixes issues working with a Perimice-201 that was experienced by a number of users.

**Comment by leroy-brooksnet** [comment]
This SMC update fixed my mouse problem.  It is the Odessa HC-3003PS (https://www.amazon.com/gp/product/B003LD0BQ6).  My Dell RT7D20 PS/2 keyboard continues to exhibit the same problem as before where the Num, Scroll, and Caps Lock LEDs all blink once when the computer is turned on, but then are off and none of the keys work.

File: x16-smc.ino [hunk 2]
```diff
@@ -3,93 +3,132 @@
...
-#define RESET_ACTIVE   LOW
-#define RESET_INACTIVE HIGH
-
+#define RESET_ACTIVE           LOW
+#define RESET_INACTIVE         HIGH
+
+// I2C
+#define I2C_ADDR               0x42 // General slave address
+
+#define CMD_POW_OFF            0x01
```
[the remark this answers is not in the document]
Comment (stefan-b-jakobsson) [reply x16-smc.ino]: So renamed

**Comment by stefan-b-jakobsson** [comment]
> This SMC update fixed my mouse problem. It is the Odessa HC-3003PS (https://www.amazon.com/gp/product/B003LD0BQ6). My Dell RT7D20 PS/2 keyboard continues to exhibit the same problem as before where the Num, Scroll, and Caps Lock LEDs all blink once when the computer is turned on, but then are off and none of the keys work.

Sounds like a problem in the keyboard init routine. If you have an external programmer for the SMC and enough patience, I could help you test a few things. Just let me know if you want to do that.

**Comment by stefan-b-jakobsson** [comment]
I found this article about the Dell RT7D20 keyboard:

https://review.coreboot.org/c/coreboot/+/486

Basically it says that differs from most keyboard, and takes longer time to initialize. Maybe we could just increase the timeout.

**Comment by leroy-brooksnet** [comment]
Thank you so much for your willingness to help.  It's very much appreciated.  I would love to help, but I don't currently have an external programmer and I don't want to brick my x16 even temporarily.

**Comment by stefan-b-jakobsson** [comment]
GitHub Actions will stop supporting Node16 during this Spring.

I have updated all available resources to use Node20, but there is still one left that is not yet ready: arduino/setup-arduino-cli@v1. They have a PR fixing that issue waiting to be merged.

**Comment by mooinglemur** [comment]
ROM side has been merged

**Comment by stefan-b-jakobsson** [comment]
I've committed changes to the keyboard and mouse init code to always wait for BAT OK in power on. Previously, the code could do a reset on power on before receiving BAT OK. Most keyboards are fine with that, but some are not, as discussed in the Discord PS/2 thread.

**Review by mooinglemur — APPROVED** [review]

**Pull request merged**